ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -89,23 +93,28 @@ public void run() {
         handle = txnHandler.getMutexAPI().acquireLock(TxnStore.MUTEX_KEY.Cleaner.name());
         startedAt = System.currentTimeMillis();
         long minOpenTxnId = txnHandler.findMinOpenTxnIdForCleaner();
+        List<CompletableFuture> cleanerList = new ArrayList<>();
         for(CompactionInfo compactionInfo : txnHandler.findReadyToClean()) {
-          clean(compactionInfo, minOpenTxnId);
+          cleanerList.add(CompletableFuture.runAsync(CompactorUtil.ThrowingRunnable.unchecked(() ->
+                  clean(compactionInfo, minOpenTxnId)), cleanerExecutor));
         }
+        CompletableFuture.allOf(cleanerList.toArray(new CompletableFuture[0])).join();
       } catch (Throwable t) {
         LOG.error("Caught an exception in the main loop of compactor cleaner, " +
-            StringUtils.stringifyException(t));
-      }
-      finally {
+                StringUtils.stringifyException(t));
+        if (cleanerExecutor != null) {
```

Comment:
what's the purpose of executor service shutdown?

Reply:
This is to handle the case when threads in executors are lost due to exception.

Reply:
You are submitting all tasks to the executor inside the loop at the end, so if there will be an exception in main loop - there won't be any running threads in executor.

Reply:
Agreed. I have removed the executor shutdown. 

But where/should we add executor shutdown? I think since the Cleaner is supposed to be running all the time. We can skip executor shutdown?

Reply:
@adesh-rao, sorry I think you have placed it originally at the right spot, at the end of run() method

Reply:
Which `InterruptedException` are you pointing at?  Also, I have moved the shutdown at the end of run method now.

Reply:
@adesh-rao, exception from (see how it's done in Initiator)           
          try {
            Thread.sleep(cleanerCheckInterval - elapsedTime);
          } catch (InterruptedException ie) {
            // What can I do about it?
          }
rest looks good to me